src/shared/condition.c

```diff
@@ -450,6 +451,87 @@ static int condition_test_architecture(Condition *c, char **env) {
         return a == b;
 }
 
+#define DTCOMPAT_FILE "/sys/firmware/devicetree/base/compatible"
+static int condition_test_firmware_devicetree_compatible(const char *dtcompatarg) {
+        int err;
+        _cleanup_free_ char *dtcompat = NULL;
+        char *compat;
+        size_t size;
+
+        err = read_full_virtual_file(DTCOMPAT_FILE, &dtcompat, &size);
+        if (err < 0) {
+                /* if the path doens't exist it is incompatible */
+                if (err != -ENOENT)
+                        log_debug_errno(errno, "Failed to open() '%s', assuming machine is incompatible: %m", DTCOMPAT_FILE);
+                return false;
+        }
+
+        /* Not sure this can happen, but play safe. */
+        if (!size) {
+                log_debug("%s has zero length, assuming machine is incompatible", DTCOMPAT_FILE);
+                return false;
+        }
+        /*
+         * /sys/firmware/devicetree/base/compatible consists of one or more
+         * strings, each ending in '\0'. So the last character in dtcompat must
+         * be a '\0'. Parsing the buffer relies on this, so check this first.
+         */
+        if (dtcompat[size - 1] != '\0') {
+                log_debug("%s is in an unknown format, assuming machine is incompatible", DTCOMPAT_FILE);
+                return false;
+        }
+        compat = dtcompat;
+        do {
+                size_t compat_len;
+
+                if (!strcmp(dtcompatarg, compat))
+                        return true;
+
+                compat_len = strlen(compat) + 1;
+
+                assert(size >= compat_len);
+
+                size -= compat_len;
+                compat += compat_len;
+        } while (size);
```

Comment:
maybe just do:

```c
l = strv_parse_nulstr(dtcompat, size);
if (!l)
        return -ENOMEM;
return strv_contains(l, dtcomparg);
```

i.e. use the parser we already have for this. yes, this will do more memory allocations, but that shouldn't really matter, this is not performance critical after all, and I much prefer if we'd reuse parsers as much as possible, since they are after all problematic for security.